### PR TITLE
Featre/implement fast path

### PR DESCRIPTION
**Pull request by HenriChabert**
Linked to [ch62283]

### CODE REVIEW

File: python-lib/dku_deeplearning_image/recipes/retrain_recipe.py
Comment (nicolasservel) [inline python-lib/dku_deeplearning_image/recipes/retrain_recipe.py:137]: `pddf` is a bit cryptic, `df` would be enough no ?

File: python-lib/dku_deeplearning_image/utils.py
Comment (nicolasservel) [inline python-lib/dku_deeplearning_image/utils.py:204]: Why split reading & preprocessing images ? They are both using custom code, and I think it would be clearer to merge them.

File: python-lib/dku_deeplearning_image/utils.py
Comment (nicolasservel) [inline python-lib/dku_deeplearning_image/utils.py:212]: why use `interleave` rather than `map` here ?

**Review by nicolasservel — CHANGES_REQUESTED** [review]
First round of reviews, did not review all tf.dataset logic yet

Got a weird error while re-training:

```
[2021/04/02-18:17:14.483] [null-err-38] [INFO] [dku.utils]  - *************** Recipe code failed **************
[2021/04/02-18:17:14.484] [null-err-38] [INFO] [dku.utils]  - Begin Python stack
[2021/04/02-18:17:14.486] [null-err-38] [INFO] [dku.utils]  - Traceback (most recent call last):
[2021/04/02-18:17:14.487] [null-err-38] [INFO] [dku.utils]  -   File "/Users/nicolasservel/tests_instances/release90/dip_home/jobs/TestKerasBinaryImageFeatCPU/Build_retrained_model__NP__2021-04-02T16-15-27.485/compute_wdYy56Ib_NP/custom-python-recipe/pyoutfkYzqUlZHPak/python-exec-wrapper.py", line 206, in <module>
[2021/04/02-18:17:14.487] [null-err-38] [INFO] [dku.utils]  -     exec(f.read())
[2021/04/02-18:17:14.487] [null-err-38] [INFO] [dku.utils]  -   File "<string>", line 45, in <module>
[2021/04/02-18:17:14.488] [null-err-38] [INFO] [dku.utils]  -   File "/Users/nicolasservel/tests_instances/release90/dip_home/plugins/installed/deeplearning-image-v2/python-lib/dku_deeplearning_image/utils.py", line 163, in wrapper
[2021/04/02-18:17:14.488] [null-err-38] [INFO] [dku.utils]  -     res = f(*args, **kwargs)
[2021/04/02-18:17:14.489] [null-err-38] [INFO] [dku.utils]  -   File "<string>", line 41, in run
[2021/04/02-18:17:14.489] [null-err-38] [INFO] [dku.utils]  -   File "/Users/nicolasservel/tests_instances/release90/dip_home/plugins/installed/deeplearning-image-v2/python-lib/dku_deeplearning_image/recipes/retrain_recipe.py", line 171, in compute
[2021/04/02-18:17:14.489] [null-err-38] [INFO] [dku.utils]  -     callback_list=callbacks
[2021/04/02-18:17:14.490] [null-err-38] [INFO] [dku.utils]  -   File "/Users/nicolasservel/tests_instances/release90/dip_home/plugins/installed/deeplearning-image-v2/python-lib/dku_deeplearning_image/recipes/retrain_recipe.py", line 116, in _retrain
[2021/04/02-18:17:14.490] [null-err-38] [INFO] [dku.utils]  -     verbose=2
[2021/04/02-18:17:14.490] [null-err-38] [INFO] [dku.utils]  -   File "/Users/nicolasservel/tests_instances/release90/dip_home/plugins/installed/deeplearning-image-v2/python-lib/dku_deeplearning_image/misc_objects/dku_model.py", line 273, in fit
[2021/04/02-18:17:14.491] [null-err-38] [INFO] [dku.utils]  -     self.model.fit(**kwargs)
[2021/04/02-18:17:14.491] [null-err-38] [INFO] [dku.utils]  -   File "/Users/nicolasservel/tests_instances/release90/dip_home/code-envs/python/plugin_deeplearning-image-v2_managed/lib/python3.6/site-packages/tensorflow/python/keras/engine/training.py", line 66, in _method_wrapper
[2021/04/02-18:17:14.492] [null-err-38] [INFO] [dku.utils]  -     return method(self, *args, **kwargs)
[2021/04/02-18:17:14.492] [null-err-38] [INFO] [dku.utils]  -   File "/Users/nicolasservel/tests_instances/release90/dip_home/code-envs/python/plugin_deeplearning-image-v2_managed/lib/python3.6/site-packages/tensorflow/python/keras/engine/training.py", line 872, in fit
[2021/04/02-18:17:14.492] [null-err-38] [INFO] [dku.utils]  -     return_dict=True)
[2021/04/02-18:17:14.493] [null-err-38] [INFO] [dku.utils]  -   File "/Users/nicolasservel/tests_instances/release90/dip_home/code-envs/python/plugin_deeplearning-image-v2_managed/lib/python3.6/site-packages/tensorflow/python/keras/engine/training.py", line 66, in _method_wrapper
[2021/04/02-18:17:14.493] [null-err-38] [INFO] [dku.utils]  -     return method(self, *args, **kwargs)
[2021/04/02-18:17:14.494] [null-err-38] [INFO] [dku.utils]  -   File "/Users/nicolasservel/tests_instances/release90/dip_home/code-envs/python/plugin_deeplearning-image-v2_managed/lib/python3.6/site-packages/tensorflow/python/keras/engine/training.py", line 1091, in evaluate
[2021/04/02-18:17:14.494] [null-err-38] [INFO] [dku.utils]  -     logs = tf_utils.to_numpy_or_python_type(logs)
[2021/04/02-18:17:14.495] [null-err-38] [INFO] [dku.utils]  - UnboundLocalError: local variable 'logs' referenced before assignment
```

Seems to come from: https://github.com/tensorflow/tensorflow/issues/38064
Did you alredy experience it ?

File: python-lib/dku_config/dss_parameter.py
Comment (HenriChabert) [reply python-lib/dku_config/dss_parameter.py:97]: Mmmh, I did that for debugging purposes. If I want to debug the plugin in a jupyter-notebook, I need to build my recipe with a valid dku_config. To do so, I print the `dku_config` when executing the recipe in dss, and I just have to copy/paste what is printed in the logs to use it in my Jupyter Notebook.
In other words, when doing a `print(self.dku_config)`, I want the printed text to be copiable, so that I just have to do `dku_config = DkuConfig(**copied_dku_config_from_log)` in my jupyter-notebook and it works.
Do you think I should rather add a function `print_as_dict()` for debugging?

File: python-lib/dku_deeplearning_image/dku_constants.py
Comment (HenriChabert) [reply python-lib/dku_deeplearning_image/dku_constants.py:75]: Actually, this constant does not exist in tensorflow-2.2. I did not want to upgrade to tensorflow-2.4 as I wanted to test in on dku11 and the CUDA version is not the right one for tf-2.4.

File: python-lib/dku_deeplearning_image/utils.py
Comment (HenriChabert) [reply python-lib/dku_deeplearning_image/utils.py:221]: It is resized in the line just under is.

Comment (nicolasservel) [reply python-lib/dku_deeplearning_image/utils.py:221]: my bad

File: python-lib/dku_deeplearning_image/utils.py
Comment (HenriChabert) [reply python-lib/dku_deeplearning_image/utils.py:131]: Isn't the temp job folder automatically cleaned? I can't find the files downloaded in the job dir

File: python-lib/dku_deeplearning_image/utils.py
Comment (HenriChabert) [reply python-lib/dku_deeplearning_image/utils.py:204]: Actually, I did that because when using a b64 image, the way of reading is different, I only need to preprocess the image. So I separated the two functions to make it possible.

File: python-lib/dku_deeplearning_image/utils.py
Comment (HenriChabert) [reply python-lib/dku_deeplearning_image/utils.py:212]: `interleave` is the function made for data extraction (https://www.tensorflow.org/guide/data_performance#parallelizing_data_extraction) and `map` for data transformation. I do not really understand the difference between these both, but I'd rather follow tf recommandations.

File: python-lib/dku_deeplearning_image/misc_objects/dku_model.py
Comment (HenriChabert) [reply python-lib/dku_deeplearning_image/misc_objects/dku_model.py]: Thanks, typo when resolving conflicts. Fixed in b17578a

File: python-lib/dku_deeplearning_image/utils.py
Comment (HenriChabert) [reply python-lib/dku_deeplearning_image/utils.py]: True. Fixed in edf74b2

File: python-lib/dku_deeplearning_image/utils.py
Comment (HenriChabert) [reply python-lib/dku_deeplearning_image/utils.py]: Fixed in 9a79b39

File: python-lib/dku_deeplearning_image/recipes/retrain_recipe.py
Comment (HenriChabert) [reply python-lib/dku_deeplearning_image/recipes/retrain_recipe.py]: Fixed in 8806f43

File: python-lib/dku_deeplearning_image/utils.py
Comment (HenriChabert) [reply python-lib/dku_deeplearning_image/utils.py]: Fixed in 9a79b39

File: python-lib/dku_deeplearning_image/utils.py
Comment (nicolasservel) [inline python-lib/dku_deeplearning_image/utils.py:221]: `read_images` does not actually read the images, but fetches them from the managed folder and yields the path to the image, so the name is confusing

File: python-lib/dku_deeplearning_image/utils.py
Comment (nicolasservel) [reply python-lib/dku_deeplearning_image/utils.py:229]: I think `interleave` usecase is more when each call builds a `Dataset` (one -> many) while `map` builds only element of the resulting dataset (one -> one).

The typical usecase for interleave is when you want to build a text Dataset out of multiple files. Each file reading will yield one dataset (where each line is an element) that you want to merge into a single big dataset for the training. 

Here you just want to map one file to one image, so I think using `map` is clearer. What do you think ?

File: python-lib/dku_deeplearning_image/utils.py
Comment (nicolasservel) [inline python-lib/dku_deeplearning_image/utils.py:242]: You are returning a tf tensor here, while if no error, you are returning a numpy array. I don't know if this works, but definitely does not look clean

File: python-lib/dku_deeplearning_image/utils.py
Comment (nicolasservel) [inline python-lib/dku_deeplearning_image/utils.py:288]: Not used ?

**Review by nicolasservel — CHANGES_REQUESTED** [review]
Another round of review

File: python-lib/dku_deeplearning_image/config_handler.py
Comment (HenriChabert) [reply python-lib/dku_deeplearning_image/config_handler.py]: You are definitely right and it was also a fault of implementation in other parameters. I rewrote dku_config a little bit to be able  to cast is once we have checked that the value is not None. See 02742f4 and the commits after that

File: python-lib/dku_deeplearning_image/utils.py
Comment (HenriChabert) [reply python-lib/dku_deeplearning_image/utils.py]: Fixed in 4f9a704

File: python-lib/dku_deeplearning_image/utils.py
Comment (HenriChabert) [reply python-lib/dku_deeplearning_image/utils.py:243]: 1. Yop, fixed in 0f2e0cd
2. True, i was a real pain to build this function, and the return value if there is an error but you are right, things should be consistant. Fixed in b6310a2

File: python-lib/dku_deeplearning_image/utils.py
Comment (HenriChabert) [reply python-lib/dku_deeplearning_image/utils.py]: Removed in 9111f63

File: python-lib/dku_deeplearning_image/utils.py
Comment (nicolasservel) [inline python-lib/dku_deeplearning_image/utils.py:131]: I think it should be the opposite, because this code would fail for unicode in py2 for instance, so check whether its `bytes`, else consider it as a string (would also work for unicode)

File: python-lib/dku_deeplearning_image/utils.py
Comment (nicolasservel) [reply python-lib/dku_deeplearning_image/utils.py:243]: I think that `preprocess_img` should return only `np.ndarray` and not `tf.tensor`. The goal of [`numpy_function`](https://www.tensorflow.org/api_docs/python/tf/numpy_function) is to abstract away the complexity of tensorflow from the used function. Tensorflow is in charge of converting the output to tensors

`preprocess_img` could be used outside of tensorflow, but currently it's not possible

**Review by nicolasservel — CHANGES_REQUESTED** [review]
`set_gpu_options` does not work as expected (found out because memory is not limited for instance) because you are comparing string values with `Enum` in a way that does not work

File: python-lib/dku_deeplearning_image/utils.py
Comment (HenriChabert) [reply python-lib/dku_deeplearning_image/utils.py:128]: I will change the level of logging to INFO after the def, but indeed, it has already been changed in a previous PR

File: python-lib/dku_deeplearning_image/utils.py
Comment (HenriChabert) [reply python-lib/dku_deeplearning_image/utils.py]: Fixed in c19a7e2. Indeed, I simplified the syntax of it to be easier to read

File: python-lib/dku_deeplearning_image/utils.py
Comment (HenriChabert) [reply python-lib/dku_deeplearning_image/utils.py]: You are definitely right, can't remember what brought me to reshape like that but I changed it in 46202c1

File: python-lib/dku_deeplearning_image/utils.py
Comment (HenriChabert) [reply python-lib/dku_deeplearning_image/utils.py]: I did the casting outside the function in 287bfe7, thanks!

File: python-lib/dku_deeplearning_image/config_handler.py
Comment (nicolasservel) [inline python-lib/dku_deeplearning_image/config_handler.py:29]: Still does not work because you compare with a string
```suggestion
        value=config.get('gpu_memory_limit') if dku_config.gpu_memory_allocation_mode == constants.GPU_MEMORY.LIMIT else 0)
```

**Review by nicolasservel — APPROVED** [review]
Except last nitpick, LGTM.

I think someone more familiar with the overall code of the plugin should have a look before merging

**Review by andreybavt — APPROVED** [review]
Looks good to me

**Pull request merged**
